File: A-169-求众数/majorityElement.go
package main

import (
	"fmt"
	"sort"
)

func majorityElement(nums []int) int {
	sort.Ints(nums)
	return nums[(len(nums) / 2)]

	//摩尔投票法
	// if len(nums) == 0 {
	// 	return 0
	// }

	// i, res := 1, nums[0]

	// for _, v := range nums[1:] {
	// 	if v != res {
	// 		i--
	// 	} else {
	// 		i++
	// 	}
	// 	fmt.Println("此时i", i)
	// 	if i <= 0 {
	// 		res = v
	// 		i = 1
	// 	}
	// }

	// return res
}

func main() {
	res := majorityElement([]int{1, 2, 2, 1, 2, 2, 1})
	fmt.Println(res)
}

File: A-169-求众数/majorityElement.py
#leetcode第169题 https://leetcode-cn.com/problems/majority-element/
def majorityElement(nums):
        itemDict = {}
        for i in range(len(nums)):
            itemDict[nums[i]] = itemDict.get(nums[i], 0)+1
        for i in range(len(nums)):
            if itemDict[nums[i]]>len(nums)//2:
                return nums[i]
            if nums[i+1]==nums[i]:
                continue
# 另一种方式
# def majorityElement(nums):
#         itemDict = {}
#         for i in range(len(nums)):
#             itemDict[nums[i]] = itemDict.get(nums[i], 0)+1
#         return max(itemDict, itemDict.get)

#第三种方式
# def majorityElement(nums):
#     nums=sorted(nums)
#     return nums[int(len(nums)/2)]



nums = [2,2,1,1,1,2,2]
print(majorityElement(nums))
